src/c_string.rs

```diff
@@ -0,0 +1,413 @@
+    /// if the bytes contain an interior nul byte (a zero byte not at the buffer's final position).
+    ///
+    /// If there's a nul byte at the end, however, the function does not fail.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    ///
+    /// let mut cstr = CString::<10>::new();
+    ///
+    /// cstr.push_bytes(b"hey").unwrap();
+    /// cstr.push_bytes(b" there\0").unwrap();
+    ///
+    /// assert_eq!(cstr.to_str(), Ok("hey there"));
+    /// ```
+    pub fn push_bytes(&mut self, bytes: &[u8]) -> Result<(), CapacityError> {
+        let Some(capacity) = self.capacity_with_bytes(bytes) else {
+            return Ok(());
+        };
+
+        if capacity > N {
+            // Can't store these bytes due to insufficient capacity.
+            return Err(CapacityError);
+        }
+
+        match memchr(0, bytes) {
+            Some(nul_pos) if nul_pos + 1 == bytes.len() => {
+                // Safety: inserted bytes are nul-terminated so
+                //         the CString will be left in a valid state
+                unsafe { self.extend_slice(bytes) }?;
+
+                Ok(())
+            }
+            Some(_nul_pos) => {
+                // Found an interior nul byte
+                Err(CapacityError)
+            }
+            None => {
+                // Given bytes had no nul byte anywhere,
+                // so we'll insert them and then add the nul byte terminator.
+
+                // We've ensured above that we have enough space left to insert these bytes,
+                // so the operations below must succeed
+
+                // Safety: CString will not have a nul terminator after this call is done,
+                //         but we'll fix that right below
+                unsafe { self.extend_slice(bytes) }.unwrap();
+
+                // Add the nul byte terminator
+                self.vec.push(0).unwrap();
+
+                Ok(())
+            }
+        }
+    }
+
+    /// Removes the nul byte terminator from the inner buffer.
+    ///
+    /// # Safety: caller must ensure to re-add the terminator after this function is called
+    #[inline]
+    unsafe fn pop_terminator(&mut self) {
+        debug_assert_eq!(self.vec.last(), Some(&0));
+
+        unsafe { self.vec.pop_unchecked() };
+    }
+
+    /// Converts a [`CString`] to a string slice without checking
+    /// that the string contains valid UTF-8.
+    ///
+    /// See the safe version, [`CStr::to_str`], for more information.
+    ///
+    /// # Safety
+    ///
+    /// The bytes passed in must be valid UTF-8.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    ///
+    /// let mut cstr = CString::<10>::new();
+    /// cstr.push_bytes(b"heapless").unwrap();
+    ///
+    /// assert_eq!(unsafe { cstr.as_str_unchecked() }, "heapless",);
+    /// ```
+    pub unsafe fn as_str_unchecked(&self) -> &str {
+        str::from_utf8_unchecked(self.inner_without_nul())
```

Comment:
Needs nested `unsafe` and a safety comment.

Reply:
I removed this associated function for now for the same reason as `len()`.